scripts/pulling_sim4life_images/pull_images.py

```diff
@@ -0,0 +1,180 @@
+        key_filename = stack["key"]
+        command = 'nvidia-smi'
+        not_found_str = 'not found'
+
+        # Create an SSH client and load the private key
+        ssh = paramiko.SSHClient()
+        ssh.set_missing_host_key_policy(paramiko.AutoAddPolicy())
+        private_key = paramiko.RSAKey.from_private_key_file(key_filename)
+
+        try:
+            # Connect to the remote host using SSH
+            print("SSH to " + host_ip)
+            ssh.connect(hostname=host_ip, port=port, username=username, pkey=private_key)
+
+            # Execute the 'nvidia-smi' command on the remote host
+            stdin, stdout, stderr = ssh.exec_command(command)
+
+            # Read the output of the command
+            output = stderr.read().decode()
+            print(output)
+
+            # Check if the string is not found in the output
+            if not_found_str not in output:
+                print("Nvidia-smi is present on this hosts. We pull the images")
+                for image in images_to_pull:
+                    print("docker pull " + stack["registry"] + "/" + image )
+                    stdin, stdout, stderr = ssh.exec_command("docker pull " + stack["registry"] + "/" + image )
+                    outputerr = stderr.read().decode()
+                    outputout = stdout.read().decode()
+                    print(outputerr)
+                    print(outputout)
+            else:
+                print("Nvidia-smi not present.")
+
+
+        except paramiko.AuthenticationException:
+            print("Authentication failed. Please verify your credentials and private key file path.")
+        except paramiko.SSHException as ssh_exception:
+            print(f"Unable to establish SSH connection. Error: {ssh_exception}")
+        finally:
+            # Close the SSH connection
+            ssh.close()
+
+
+# This function list aws autoscaled instance that needs to pull the image, either for aws-staging or aws-production, based on their label.
+# It returns a list of the publics ips of theses instances
+def list_aws_instances(stack_arg):
+    if stack_arg == "aws-production":
+        message = "autoscaling-osparc-production"
+    if stack_arg == "aws-staging":
+        message = "autoscaling-osparc-staging"
+
+    ec2 = boto3.client('ec2', aws_access_key_id=os.getenv('AWS_ACCESS_KEY'), aws_secret_access_key=os.getenv('AWS_SECRET_KEY'), region_name="us-east-1")
+    response = ec2.describe_instances(Filters=[{'Name': 'tag:Name', 'Values': [message]}])
+    # Extract the public IP addresses of each instance 
+    public_ips = []
+    for reservation in response['Reservations']:
+        for instance in reservation['Instances']:
+            if "PublicIpAddress" in instance:
+                public_ips.append(instance['PublicIpAddress'])
+
+    return public_ips
+
+
+
+def clone_repos():
+    # Set up the GitLab API client
+    gl = gitlab.Gitlab(os.getenv('GITLAB_URL'), private_token=os.getenv('GITLAB_ACCESS_TOKEN'))
```

Comment:
Is this your personal access token usually or do we have a devops account? I think we shall have a devops account if we dont have it yet, I can take care of it if you want :)